clang/docs/checkedc/static-analyzer-with-z3.md

```diff
@@ -0,0 +1,98 @@
+
+However the analyzer is not aware of the additional annotations that Checked-C maintains for bounds checking. Furthermore, the memory bounds checking checkers are in alpha and cannot detect many simple memory access violation bugs.
+
+This document discusses the implementation of a clang static analyzer checker (`SimpleBounds`) that improves the memory bounds checking of `ArrayBound` checkers, and also demonstrate the use of theorem provers (e.g. Z3) to analyze the bounds expressions.
+
+We have compiled and tested our system on Linux, so currently installation instructions are for Linux environments.
+
+## Compiling Checked-C with Z3
+
+We use Z3's C++ API to query Z3 inside our checker. Therefore, we need to compile Checked-C with the Z3 support, and for that we first need to install Z3. The following steps fetches, compiles and install Z3. For more details refer to [Z3](https://github.com/Z3Prover/z3/) github page.
+
+1. `git clone https://github.com/Z3Prover/z3/`
+2. `cd z3`
+3. `mkdir build`
+4. `cd build`
+5. `cmake -G "Unix Makefiles" ../`
+6. `make -j8`
+7. `sudo make install`
+
+Next, we can build the Checked-C compiler by following the instructions in [Setup-and-Build](Setup-and-Build.md), with the only difference that we should pass this flag when running cmake: `-DCLANG_ANALYZER_ENABLE_Z3_SOLVER=ON`.
+
+
+## Using Z3 in Clang
+
+When compiled with Z3, Clang static analyzer can leverage Z3 to prune some of the false positive bug reports generated by its checkers by improving upon the solving ability of its internal solver (`RangeConstraintManager`). When a bug report is generated, the path from the input to that bug point will be encoded as a SMT formula and passed to Z3. Clang basically asks Z3 for reachability of the bug point. If the bug is unreachable, the report is suppressed.
+
+To analyze any code the following command is used:
+```
+clang -cc1 -analyze main.c
+```
+
+You can see the list of available checkers by running: `clang -cc1 -analyzer-checker-help`.
+Some of the checkers are disabled by default, to turn on any checker, you can run:
+```
+clang -cc1 -analyze -analyzer-checker {checker1,checker2} main.c
+```
+
+To enable the Z3 bug refutation, you should specify its flag in the analyzer config:
+```
+clang -cc1 -analyze -analyzer-checker {checker1,checker2} -analyzer-config 'crosscheck-with-z3=true' main.c
+```
+
+## Our Checker
+
+Our checker, called `SimpleBounds` and categorized under `alpha.Security`, is written with the primary goal of checking whether the memory accesses within an unchecked code are following the bounds-safe interface specification. The main two points in the implementation of this checker are:
+
+1. Taking Checked-C's bounds information into account
+2. Using Z3 for complex index and bounds expressions
+
+The file that implements this checker is `lib/StaticAnalyzer/Checkers/SimpleBoundsChecker.cpp`.
+
+An example code snippet that clang's internal solver can not handle is:
+```c

```

Comment:
I think there is extra c character after the quotes.

Reply:
It is for highlighting C codes.